mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.cpp

```diff
@@ -114,68 +125,1196 @@ class LooseCompressedLevel : public SparseLevel {
+                                ValueRange{tupleId, C_IDX(i)});
+    }
+  }
+
+  SmallVector<Value> loadItVals(OpBuilder &b, Location l, Value tupleId) const {
+    SmallVector<Value> ret;
+    for (unsigned i = 0; i < tupleSz; i++) {
+      Value v = b.create<memref::LoadOp>(l, subSectPosBuf,
+                                         ValueRange{tupleId, C_IDX(i)});
+      ret.push_back(v);
+    }
+    return ret;
+  }
+
+  bool isSubSectRoot() const {
+    return !parent || !llvm::isa<NonEmptySubSectIterator>(parent);
+  }
+
+  // Generate code that inflate the current subsection tree till the current
+  // level such that every leaf node is visited.
+  ValueRange inflateSubSectTree(OpBuilder &b, Location l, ValueRange reduc,
+                                TraverseBuilder builder) const;
+
+  bool randomAccessible() const override {
+    return delegate->randomAccessible();
+  };
+  bool iteratableByFor() const override { return randomAccessible(); };
+  Value upperBound(OpBuilder &b, Location l) const override {
+    auto *p = dyn_cast_or_null<NonEmptySubSectIterator>(parent);
+    Value parentUB =
+        p && p->lvl == lvl ? p->upperBound(b, l) : delegate->upperBound(b, l);
+    return ADDI(SUBI(parentUB, subSectSz), C_IDX(1));
+  };
+
+  void genInit(OpBuilder &b, Location l, const SparseIterator *) override;
+
+  void locate(OpBuilder &b, Location l, Value crd) override {
+    Value absOff = crd;
+
+    if (isSubSectRoot())
+      delegate->locate(b, l, absOff);
+    else
+      assert(parent->lvl + 1 == lvl);
+
+    seek(ValueRange{absOff, absOff, C_TRUE});
+    updateCrd(crd);
+  }
+
+  Value toSubSectCrd(OpBuilder &b, Location l, Value wrapCrd) const {
+    assert(stride == 1);
+    return SUBI(wrapCrd, getAbsOff());
   }
+
+  Value genNotEnd(OpBuilder &b, Location l) override { return getNotEnd(); };
+
+  Value deref(OpBuilder &b, Location l) override {
+    // Use the relative offset to coiterate.
+    Value crd;
+    auto *p = dyn_cast_or_null<NonEmptySubSectIterator>(parent);
+    if (p && p->lvl == lvl)
+      crd = SUBI(getAbsOff(), p->getAbsOff());
+    crd = getAbsOff();
+
+    updateCrd(crd);
+    return crd;
+  };
+
+  ValueRange forward(OpBuilder &b, Location l) override;
+
+  Value getMinCrd() const { return subSectMeta[0]; }
+  Value getAbsOff() const { return subSectMeta[1]; }
+  Value getNotEnd() const { return subSectMeta[2]; }
+
+  // Number of values required to serialize the wrapped iterator.
+  unsigned tupleSz;
+  // Max number of tuples, and the actual number of tuple.
+  Value maxTupleCnt, tupleCnt;
+  // The memory used to cache the tuple serialized from the wrapped iterator.
+  Value subSectPosBuf;
+
+  const Value subSectSz;
+  const unsigned stride;
+
+  const SparseIterator *parent;
+  std::unique_ptr<SparseIterator> delegate;
+
+  Value subSectMeta[3]; // minCrd, absolute offset, notEnd
+};
+
+class SubSectIterator;
+
+// A simple helper that helps generating code to traverse a subsection, used
```

Comment:
remove "simple"

also, a struct is not really a helper by itself, so perhaps make that a bit more clear